packages/keystone/lib/ListTypes/hooks.js

```diff
@@ -0,0 +1,166 @@
+    resolvedData = omitBy(resolvedData, key => typeof resolvedData[key] === 'undefined');
+
+    // Run the schema-level field hooks, passing in the results from the field
+    // type hooks
+    resolvedData = {
+      ...resolvedData,
+      ...(await mapToFields(
+        this.fields.filter(field => field.hooks.resolveInput),
+        field => field.hooks.resolveInput({ ...args, resolvedData })
+      )),
+    };
+
+    // And filter out the `undefined`s again.
+    resolvedData = omitBy(resolvedData, key => typeof resolvedData[key] === 'undefined');
+
+    if (this.hooks.resolveInput) {
+      // And run any list-level hook
+      resolvedData = await this.hooks.resolveInput({ ...args, resolvedData });
+      if (typeof resolvedData !== 'object') {
+        throw new Error(
+          `Expected ${
+            this.listKey
+          }.hooks.resolveInput() to return an object, but got a ${typeof resolvedData}: ${resolvedData}`
+        );
+      }
+    }
+
+    // Finally returning the amalgamated result of all the hooks.
+    return resolvedData;
+  }
+
+  async validateInput({ resolvedData, existingItem, context, operation, originalInput }) {
+    const args = { resolvedData, existingItem, context, originalInput, operation };
+    // Check for isRequired
+    const fieldValidationErrors = this.fields
+      .filter(
+        field =>
+          field.isRequired &&
+          !field.isRelationship &&
+          ((operation === 'create' &&
+            (resolvedData[field.path] === undefined || resolvedData[field.path] === null)) ||
+            (operation === 'update' &&
+              Object.prototype.hasOwnProperty.call(resolvedData, field.path) &&
+              (resolvedData[field.path] === undefined || resolvedData[field.path] === null)))
+      )
+      .map(f => ({
+        msg: `Required field "${f.path}" is null or undefined.`,
+        data: { resolvedData, operation, originalInput },
+        internalData: {},
+      }));
+    if (fieldValidationErrors.length) {
+      this._throwValidationFailure({ errors: fieldValidationErrors, operation, originalInput });
+    }
+
+    const fields = this._fieldsFromObject(resolvedData);
+    await this._validateHook({ args, fields, operation, hookName: 'validateInput' });
+  }
+
+  async validateDelete({ existingItem, context, operation }) {
+    const args = { existingItem, context, operation };
+    const fields = this.fields;
+    await this._validateHook({ args, fields, operation, hookName: 'validateDelete' });
+  }
+
+  async _validateHook({ args, fields, operation, hookName }) {
+    const { originalInput } = args;
+    const fieldValidationErrors = [];
+    // FIXME: Can we do this in a way where we simply return validation errors instead?
+    args.addFieldValidationError = (msg, _data = {}, internalData = {}) =>
+      fieldValidationErrors.push({ msg, data: _data, internalData });
+    await mapToFields(fields, field => field[hookName](args));
+    await mapToFields(
```

Comment:
Do we need two `await` calls here?

Reply:
Yep, we want to ensure that the hooks defined by the field type itself are executed before any user defined hooks for the same field.